src/OrleansRuntime/Facet/GrainConstructorFacetArguementsFactory.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans.Runtime
+{
+    internal class GrainConstructorFacetArguementsFactory
+    {
+        private static readonly Type facetFactoryType = typeof(IParameterFacetFactory<>);
+        /// <summary>
+        /// Cached constructor arguement factorys by type
+        /// TODO: consider storing in grain type data and constructing at startup to avoid runtime errors. - jbragg
+        /// </summary>
+        private readonly ConcurrentDictionary<Type, ArguementsFactory> arguementsFactorys;
+        private readonly IServiceProvider services;
+
+        public GrainConstructorFacetArguementsFactory(IServiceProvider services)
+        {
+            this.services = services;
+            arguementsFactorys = new ConcurrentDictionary<Type, ArguementsFactory>();
+        }
+
+        public Type[] ArguementTypes(Type type)
+        {
+            ArguementsFactory arguementsFactory = arguementsFactorys.GetOrAdd(type, t => new ArguementsFactory(this.services, t));
+            return arguementsFactory.ArguementTypes;
+        }
+
+        public object[] CreateArguements(IGrainActivationContext grainActivationContext)
+        {
+            ArguementsFactory arguementsFactory = arguementsFactorys.GetOrAdd(grainActivationContext.GrainType, type => new ArguementsFactory(this.services, type));
+            return arguementsFactory.CreateArguements(grainActivationContext);
+        }
+
+        /// <summary>
+        /// Facet Arguement factory
+        /// </summary>
+        private class ArguementsFactory
+        {
+            private readonly List<Factory<IGrainActivationContext, object>> arguementFactorys;
+
+            public ArguementsFactory(IServiceProvider services, Type type)
+            {
+                this.arguementFactorys = new List<Factory<IGrainActivationContext, object>>();
+                List<Type> types = new List<Type>();
+                IEnumerable<ParameterInfo> parameters = type.GetConstructors()
+                                                            .FirstOrDefault()?
+                                                            .GetParameters() ?? Enumerable.Empty<ParameterInfo>();
+                foreach (ParameterInfo parameter in parameters)
+                {
+                    var attribute = parameter.GetCustomAttribute<FacetAttribute>();
+                    if (attribute == null) continue;
+                    IFactoryBridge bridge = Activator.CreateInstance(typeof(FactoryBridge<>).MakeGenericType(attribute.GetType())) as IFactoryBridge;
+                    if (bridge == null) continue;
+                    bridge.Factory = services.GetRequiredService(facetFactoryType.MakeGenericType(attribute.GetType()));
```

Comment:
It took a long time to understand what this code is doing, maybe you can check my understanding:
* Some attribute, `TAttribute : FacetAttribute` on the parameter holds the config.
* Get the `factoryFactory` from the container via the type `IParameterFacetFactory<TAttribute>`.
* Call `factoryFactory.Create(parameterInfo, attribute)` to get `factory`.
* On activation, call `factory(context, services)` to get the value to pass to the grain constructor.
* Under the hood, `factory` will pull the strongly typed factory interface (eg `INamedStorageFeatureFactory`) from the container and call it using some value captured in the closure (eg the config specified on the attribute).